tools/eslint-plugin/dist/rules/public-extension-exports.js

```diff
@@ -0,0 +1,229 @@
+      names.forEach(name => {
+        const kind = getSyntaxKindFriendlyName(declaration.kind);
+        const csvString = createCsvString(name, kind);
+        fs.writeFileSync(apiFilePath, csvString, { flag: "a" });
+      });
+    }
+
+    function getParentSymbolName(declaration) {
+      if (declaration.parent && declaration.parent.symbol && !declaration.parent.symbol.escapedName.startsWith('"'))
+        return declaration.parent.symbol.escapedName;
+      return undefined;
+    }
+
+    // reports an error if namespace doesn't have a valid @extensions tag but a member does
+    function checkNamespaceTags(declaration, node) {
+      const tags = ts.getJSDocTags(declaration.parent);
+      if (!tags || tags.length === 0)
+        return;
+
+      for (const tag of tags) {
+        if (tag.tagName.escapedText === extensionsTag) {
+          return;
+        }
+      }
+      context.report({
+        node,
+        messageId: "namespace",
+        data: {
+          name: ts.getNameOfDeclaration(declaration.parent)?.getFullText(),
+        }
+      });
+    }
+
+    // returns true if it was added to the API without error
+    function checkJsDoc(declaration, node) {
+      if (!declaration || !declaration.jsDoc)
+        return;
+
+      const tags = ts.getJSDocTags(declaration);
+      if (!tags || tags.length === 0)
+        return;
+
+      let jsDocExtensionTag = tags.find(tag => tag?.tagName?.escapedText === extensionsTag);
+      let jsDocPreviewTag = tags.find(tag => tag?.tagName?.escapedText === previewTag);
+      // Has extension API tag
+      if (jsDocExtensionTag) {
+        addToApiList(declaration, jsDocPreviewTag);
+        const validReleaseTag = tags.some(tag => releaseTags.includes(tag?.tagName?.escapedText));
+        if (validReleaseTag) {
+          return true;
+        } else {
+          let name;
+          if (declaration.kind === ts.SyntaxKind.Constructor)
+            name = declaration.parent?.symbol?.escapedName;
+          else {
+            name = declaration.symbol?.escapedName;
+            const parentSymbol = getParentSymbolName(declaration);
+            if (parentSymbol)
+              name = `${parentSymbol}.${name}`;
+          }
+
+          context.report({
+            node,
+            messageId: "forbidden",
+            data: {
+              kind: getSyntaxKindFriendlyName(declaration.kind),
+              name,
+              releaseTags: releaseTags,
+            }
+          });
+        }
+      }
+    }
+
+    function isNamespace(declaration) {
+      return ts.isModuleBlock(declaration) && ts.isModuleDeclaration(declaration.parent);
+    }
+
+    function check(declaration, node) {
+      if (!declaration)
+        return;
+      if (checkJsDoc(declaration, node)) {
+        if (declaration.parent && isNamespace(declaration.parent))
+          checkNamespaceTags(declaration.parent, node);
+      }
+    }
+
+    return {
+      TSNamespaceExportDeclaration(node) {
```

Comment:
All these functions are identical and I recently learned you can use custom selectors for function names here:
https://eslint.org/docs/developer-guide/selectors#listening-for-selectors-in-rules
You should be able to put these node types into a comma-separated string and use that as the function name.

That's just an idea, though, so I'll resolve this comment immediately so it doesn't block the PR.